### PR TITLE
Add LabelEdit to TreeView

### LINKED ISSUE #9956: Label edit in TreeView has incorrect position in accessibility tree
### .NET version

.NET 8.0.100-rc.2.23469.4

### Did it work in .NET Framework?

No

### Did it work in any of the earlier releases of .NET Core or .NET 5+?

No, this issue also repro's in .NET3.1/5.0/6.0/7.0.

### Issue description

This issue is similar with [7263](https://github.com/dotnet/winforms/issues/7263). Edit control of TreeViewNode will be placed as a first child of the TreeView in accessibility tree.

before
before2

### Steps to reproduce

1. Create a Winforms .NET Core application
2. Add a TreeView control
3. Add some root node and child node
4. Set TreeView control LabelEdit property is true
5. Build and run
6. Open Inspect or Accessibility Insights tool 
7. Click twice on any tree node to start editing it and find Edit control place in tools

### PR DESCRIPTION
**Pull request by SimonZhao888**
Fixes #9956 

## Proposed changes

- Adding a shared control of LabelEdit for ListView and TreeView .
- Add LabelEdit to the TreeView.
- Update LabelEdit of ListView.
- Add test cases for LabelEdit of TreeView.

## Customer Impact

- Label edit in TreeView has correct position in accessibility tree.

## Regression? 

- No

## Risk

- Minimal

### Before
The LabelEdit control of TreeViewNode will be placed as a first child of the TreeView in accessibility tree.
image

### After
The LabelEdit control of TreeViewNode will place as  a child of the Editing TreeNode in accessibility tree. 
image

## Test methodology 

- Manual (Insights, Inspect)

## Test environment(s) 

- 9.0.100-alpha.1.23504.14

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/winforms/pull/10074)

### CODE REVIEW

File: src/System.Windows.Forms/src/System/Windows/Forms/TreeView.cs [hunk 1]
```diff
@@ -3020,10 +3038,10 @@ private unsafe void WmNotify(ref Message m)
                     TvnBeginDrag(MouseButtons.Right, nmtv);
                     break;
                 case PInvoke.TVN_BEGINLABELEDITW:
-                    m.ResultInternal = (LRESULT)TvnBeginLabelEdit(*(NMTVDISPINFOW*)(nint)m.LParamInternal);
+                    TvnBeginLabelEdit(*(NMTVDISPINFOW*)(nint)m.LParamInternal, out m.ResultInternal);
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/src/System/Windows/Forms/TreeView.cs:3041]: What was the reason to change this function signature? This is a private one, so we can change the signature from returning a value to a void with an out parameter, but return value is more explicit and is easier to read.

File: src/System.Windows.Forms/src/System/Windows/Forms/TreeViewLabelEditAccessibleObject.cs [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+using Windows.Win32.System.Com;
+using Windows.Win32.UI.Accessibility;
+using UiaCore = Interop.UiaCore;
+
+namespace System.Windows.Forms;
+
+internal unsafe class TreeViewLabelEditAccessibleObject : AccessibleObject
+{
+    private const string TREE_VIEW_LABEL_EDIT_AUTOMATION_ID = "1";
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/src/System/Windows/Forms/TreeViewLabelEditAccessibleObject.cs:12]: Where does this value come from?

File: src/System.Windows.Forms/src/System/Windows/Forms/TreeViewLabelEditAccessibleObject.cs [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+    private readonly WeakReference<TreeViewLabelEditNativeWindow> _labelEdit;
+    private readonly TreeViewLabelEditUiaTextProvider _textProvider;
+    private int[]? _runtimeId;
+
+    public TreeViewLabelEditAccessibleObject(TreeView owningTreeView, TreeViewLabelEditNativeWindow labelEdit)
+    {
+        _owningTreeView = owningTreeView.OrThrowIfNull();
+        _labelEdit = new(labelEdit);
+        UseStdAccessibleObjects(labelEdit.Handle);
+        _textProvider = new TreeViewLabelEditUiaTextProvider(owningTreeView, labelEdit, this);
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/src/System/Windows/Forms/TreeViewLabelEditAccessibleObject.cs:24]:
```suggestion
        _textProvider = new (owningTreeView, labelEdit, this);
```

File: src/System.Windows.Forms/src/System/Windows/Forms/TreeViewLabelEditAccessibleObject.cs
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/src/System/Windows/Forms/TreeViewLabelEditAccessibleObject.cs:1]: Is there any way to share code between the ListView and TreeView edit control instances?

Comment (SimonZhao888) [reply src/System.Windows.Forms/src/System/Windows/Forms/TreeViewLabelEditAccessibleObject.cs:1]: I'm working on it now.

File: src/System.Windows.Forms/src/System/Windows/Forms/TreeViewLabelEditAccessibleObject.cs (hunk 2 above)
Comment (SimonZhao888) [reply src/System.Windows.Forms/src/System/Windows/Forms/TreeViewLabelEditAccessibleObject.cs:12]: The value of AutomationID does not seem to have any special meaning.
Is it more appropriate to set AutomationId to owner.name?

File: src/System.Windows.Forms/src/System/Windows/Forms/TreeViewLabelEditAccessibleObject.cs
Comment (Tanya-Solyanik) [reply src/System.Windows.Forms/src/System/Windows/Forms/TreeViewLabelEditAccessibleObject.cs:1]: I'll put this PR in a draft mode then.

Comment (SimonZhao888) [reply src/System.Windows.Forms/src/System/Windows/Forms/TreeViewLabelEditAccessibleObject.cs:1]: Done.

File: src/System.Windows.Forms/src/System/Windows/Forms/ListViewLabelEditAccessibleObject.cs [hunk 4]
```diff
@@ -1,31 +1,29 @@
...
 using UiaCore = Interop.UiaCore;
+
 namespace System.Windows.Forms;
 
-internal unsafe class ListViewLabelEditAccessibleObject : AccessibleObject
+internal unsafe class ListViewLabelEditAccessibleObject : LabelEditAccessibleObject
 {
     private const string LIST_VIEW_LABEL_EDIT_AUTOMATION_ID = "1";
 
     private readonly ListView _owningListView;
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/src/System/Windows/Forms/ListViewLabelEditAccessibleObject.cs:13]: Is it possible to get rid of this strong reference? If this is not an easy change, we can address it later.
Also please test for leaks in this scenario.

File: src/System.Windows.Forms/src/System/Windows/Forms/LabelEditUiaTextProvider.cs [hunk 5]
```diff
@@ -5,31 +5,31 @@
...
-    ///  Since the label edit inside the ListView is always single-line, for optimization
+    ///  Since the label edit inside the TreeView/ListView is always single-line, for optimization
     ///  we always return 1 as the number of lines
     /// </summary>
     private const int OwnerChildEditLinesCount = 1;
 
     private readonly IHandle<HWND> _owningChildEdit;
     private readonly AccessibleObject _owningChildEditAccessibilityObject;
-    private readonly ListView _owningListView;
+    private readonly Control _owningControl;
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/src/System/Windows/Forms/LabelEditUiaTextProvider.cs:28]: Is it possible to get rid of this strong reference?

**Review by Tanya-Solyanik — COMMENTED** [review]
Looks great, thank you! I added some small suggestions

**Pull request closed without merging**